### DISCUSSION TITLE
Proposal: yield each operator

### DISCUSSION THREAD

**Discussion by aeroelec** [dotnet/csharplang#8949] [Language Ideas]
Consider adding yield each operator that iterates through a list/array/enumerable, yield returning each element.

**Option 1**
The syntax could be
```
yield each var item in list where ... ;
```

This would simplify code to iterate through elements of a list. For example,
```
foreach (int i in numbers)
{
    if (i > 300) yield return i;
}
```
could be simplified to
```
yield each int i in numbers where i > 300;
```

If where clause is not needed, discard could be used
```
yield each _ in list;
```

**Option 2**
The syntax could be
```
yield each list;
```

Using the same example as in option 1,
```
foreach (int i in numbers)
{
    if (i > 300) yield return i;
}
```
could be simplified to
```
yield each numbers.Where(i => i > 300);
```

This is also useful for recursive iterators, such as
```
public static IEnumerable<Control> Children(this Control control)
{
    foreach (Control child in control.Controls)
    {
        yield return child;
        foreach (var grandchild in Children(child))
        {
            yield return grandchild;
        }
    }
}
```
could be simplified to
```
public static IEnumerable<Control> Children(this Control control)
{
    foreach (Control child in control.Controls)
    {
        yield return child;
        yield each Children(child);
    }
}
```

**Comment by 333fred** [comment]
Duplicate of https://github.com/dotnet/csharplang/discussions/378.
